extra/plugins/dioptra_optic/models.py

```diff
@@ -0,0 +1,465 @@
+# This Software (Dioptra) is being made available as a public service by the
+# National Institute of Standards and Technology (NIST), an Agency of the United
+# States Department of Commerce. This software was developed in part by employees of
+# NIST and in part by NIST contractors. Copyright in portions of this software that
+# were developed by NIST contractors has been licensed or assigned to NIST. Pursuant
+# to Title 17 United States Code Section 105, works of NIST employees are not
+# subject to copyright protection in the United States. However, NIST may hold
+# international copyright in software created by its employees and domestic
+# copyright (or licensing rights) in portions of software that were assigned or
+# licensed to NIST. To the extent that NIST holds copyright in this software, it is
+# being made available under the Creative Commons Attribution 4.0 International
+# license (CC BY 4.0). The disclaimers of the CC BY 4.0 license apply to all parts
+# of the software developed or licensed by NIST.
+#
+# ACCESS THE FULL CC BY 4.0 LICENSE HERE:
+# https://creativecommons.org/licenses/by/4.0/legalcode
+from types import ModuleType
+from typing import Any, Callable
+
+import keras
+import keras.applications as kapp
+import numpy as np
+import pandas as pd
+import structlog
+import tensorflow as tf
+from keras.layers import (
+    BatchNormalization,
+    Conv2D,
+    Dense,
+    Dropout,
+    Flatten,
+    Input,
+    MaxPooling2D,
+)
+from keras.models import Sequential
+
+from dioptra import pyplugs
+
+from .utils import DioptraMetricsLoggingCallback
+
+LOGGER = structlog.get_logger()
+
+
+@pyplugs.register
+def create_model(
+    model_name: str,
+    dataset: tf.data.Dataset,
```

Comment:
Design-wise it's surprising that initializing a model architecture requires having a tf.data.Dataset object available. It's a heavy dependency for a function like this when all you actually need here are the _specifications_ that set the dimensions for the input and output layers.

An idea is to modify the `data.load_dataset_from_tfds` and `load_dataset_from_image_directory` plugin tasks and have them return a dataclass that contains a metadata spec for the dataset, for classification that's the input_shape and the number of classes. Then you could pass this metadata spec object to this function instead of the full dataset. This would be also be nice for testing purposes.

Reply:
Yeah, that's what I would suggest. Based on the return signature, it'd be something like:

```python
tuple[tf.data.Dataset | None, tf.data.Dataset | None, tf.data.Dataset | None, DatasetMetadataClass]
```

And I agree, this should be a separate plugin parameter type in the task engine.